### PR TITLE
Reland: implement build bundle with assemble

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

This is a reland of https://github.com/flutter/flutter/pull/37508. The depfile's were not correctly escaped for windows, which caused failures when rebuilding.

Additionally fixes a bug in the file cache cleanup, and ensures we use a package uri when building a snapshot in assemble.

### CODE REVIEW

**Review by jmagman — APPROVED** [review]
Last 3 commits since the master merge LGTM

File: packages/flutter_tools/lib/src/build_system/targets/dart.dart [hunk 1]
```diff
@@ -55,6 +60,111 @@ List<File> listDartSources(Environment environment) {
...
+
+    final AssetBundle assetBundle = AssetBundleFactory.instance.createBundle();
+    await assetBundle.build();
+    final Pool pool = Pool(64);
+    await Future.wait<void>(
+      assetBundle.entries.entries.map<Future<void>>((MapEntry<String, DevFSContent> entry) async {
+        final PoolResource resource = await pool.request();
+        try {
+          final File file = fs.file(fs.path.join(environment.outputDir.path, entry.key));
+          file.parent.createSync(recursive: true);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/dart.dart]: What happens when one of these file operations fails?

File: packages/flutter_tools/lib/src/bundle.dart [hunk 2]
```diff
@@ -124,6 +140,66 @@ class BundleBuilder {
...
+  final File depfile = fs.file(depfilePath);
+  if (!depfile.parent.existsSync()) {
+    depfile.parent.createSync(recursive: true);
+  }
+  depfile.writeAsStringSync(buffer.toString());
+}
+
+void _writeFilesToBuffer(List<File> files, StringBuffer buffer) {
+   for (File outputFile in files) {
+    if (platform.isWindows) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/bundle.dart]: nit: 2 space indent

**Review by zanderso — APPROVED** [review]
lgtm w/ question and nit.

File: packages/flutter_tools/lib/src/build_system/targets/dart.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/dart.dart]: We should throw, will update

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/dart.dart]: actually, we're already propagating the future, just ensuring that the resource is released first

File: packages/flutter_tools/lib/src/bundle.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/bundle.dart]: Fixed

**Comment by jonahwilliams** [comment]
I can't see if the devicelab failure is real or not, so I'll rollback for now

**Pull request merged**